test/pytest/test_bridge.py

```diff
@@ -0,0 +1,114 @@
+import argparse
+import asyncio
+import json
+import sys
+import unittest
+
+from typing import Any, Dict, Tuple
+
+import systemd_ctypes
+from cockpit.router import Router
+
+asyncio.set_event_loop_policy(systemd_ctypes.EventLoopPolicy())
+
+
+class MockTransport(asyncio.Transport):
+    queue: asyncio.Queue[Tuple[str, bytes]]
+
+    def send_json(self, _channel: str, **kwargs) -> None:
+        self.send_data(_channel, json.dumps(kwargs).encode('ascii'))
+
+    def send_data(self, channel: str, data: bytes) -> None:
+        msg = channel.encode('ascii') + b'\n' + data
+        msg = str(len(msg)).encode('ascii') + b'\n' + msg
+        self.protocol.data_received(msg)
+
+    def send_init(self, version=1, host='test', **kwargs):
+        self.send_json('', command='init', version=version, host=host, **kwargs)
+
+    def send_open(self, channel, payload, **kwargs):
+        self.send_json('', command='open', channel=channel, payload=payload, **kwargs)
+
+    def send_done(self, channel, **kwargs):
+        self.send_json('', command='done', channel=channel, **kwargs)
+
+    def send_close(self, channel, **kwargs):
+        self.send_json('', command='close', channel=channel, **kwargs)
+
+    def send_ping(self, **kwargs):
+        self.send_json('', command='ping', **kwargs)
+
+    def __init__(self, protocol: asyncio.Protocol):
+        self.queue = asyncio.Queue()
+        self.protocol = protocol
+        protocol.connection_made(self)
+
+    def write(self, data: bytes) -> None:
+        # We know that the bridge only ever writes full frames at once, so we
+        # can disassemble them immediately.
+        _, channel, data = data.split(b'\n', 2)
+        self.queue.put_nowait((channel.decode('ascii'), data))
+
+    def close(self) -> None:
+        pass
+
+    async def next_frame(self) -> Tuple[str, bytes]:
+        return await self.queue.get()
+
+    async def next_msg(self, expected_channel) -> Dict[str, Any]:
+        channel, data = await self.next_frame()
+        assert channel == expected_channel
+        return json.loads(data)
+
+    async def assert_data(self, expected_channel: str, expected_data: bytes) -> None:
+        channel, data = await self.next_frame()
+        assert channel == expected_channel
+        assert data == expected_data
+
+    async def assert_msg(self, expected_channel, **kwargs) -> None:
+        msg = await self.next_msg(expected_channel)
+        assert msg == msg | {k.replace('_', '-'): v for k, v in kwargs.items()}
+
+
+class TestBridge(unittest.IsolatedAsyncioTestCase):
+    async def start(self) -> Tuple[Router, MockTransport]:
+        bridge = Router()
+        transport = MockTransport(bridge)
+
+        await transport.assert_msg('', command='init')
+        transport.send_init()
+
+        return bridge, transport
+
+    async def test_echo(self):
+        bridge, transport = await self.start()
+
+        transport.send_open('1', 'echo')
```

Comment:
That `'1'` here is irritating. See my question above why `send_open()` even expects a channel name.

Reply:
The client always decides the names of channels.  We really do need to pick a channel name and send it.

Perhaps we could keep an internal counter on the MockTransport object for this purpose, but then we'd need to return it and check it against subsequent fields...